test/Model/UserSpec.hs

```diff
@@ -14,61 +14,101 @@ main = hspec spec
+                        , credsIdent = "1"
+                        , credsExtra =
+                            [ ("name", "foo")
+                            , ("email", "bar@gmail.com")
+                            ]
+                        }
 
-        it "authenticates known users even if profile data is missing" $ do
-            Entity userId user <- runDB $ createUser "1"
-            let creds = Creds
-                    { credsPlugin = userPlugin user
-                    , credsIdent = userIdent user
-                    , credsExtra = []
-                    }
+                Right userId <- tryAuthentication creds
 
-            (runDB $ authenticateUser creds) `shouldReturn` Just userId
+                Just user <- runDB $ DB.get userId
+                userName user `shouldBe` "foo"
+                userEmail user `shouldBe` "bar@gmail.com"
 
-        it "does not authenticate unknown users if profile data is missing" $ do
-            let creds = Creds
-                    { credsPlugin = "github"
-                    , credsIdent = "1"
-                    , credsExtra = []
-                    }
+            it "updates an existing user's profile data" $ do
+                let creds = Creds
+                        { credsPlugin = "github"
+                        , credsIdent = "1"
+                        , credsExtra =
+                            [ ("name", "foo")
+                            , ("email", "bar@gmail.com")
+                            ]
+                        }
+                userId <- runDB $ do
+                    Entity planId _ <- createFreePlan
+
+                    insert $ User
+                        { userName = ""
+                        , userEmail = ""
+                        , userPlugin = credsPlugin creds
+                        , userIdent = credsIdent creds
+                        , userPlan = planId
+                        , userStripeId = Nothing
+                        }
+
+                tryAuthentication creds `shouldReturn` Right userId
+
+                Just user <- runDB $ DB.get userId
+                userName user `shouldBe` "foo"
+                userEmail user `shouldBe` "bar@gmail.com"
+
+            it "errors if name is missing" $ do
+                let creds = Creds
+                        { credsPlugin = "github"
+                        , credsIdent = "1"
+                        , credsExtra = [("email", "foo@gmail.com")]
+                        }
+
+                tryAuthentication creds `shouldReturn` Left "missing key name"
+
+            it "errors if email is missing" $ do
+                let creds = Creds
+                        { credsPlugin = "github"
+                        , credsIdent = "1"
+                        , credsExtra = [("name", "foo")]
+                        }
+
+                tryAuthentication creds `shouldReturn` Left "missing key email"
+
+-- Helper required to disambiguate AuthId m ~ UserId type
```

Comment:
This may have been a mistake on my part, I'll check.

In the meantime, I can explain a little...

It has to do with type families: `AuthId m ~ UserId` means `AuthId m` and `UserId` are in a type family. It's how `Yesod.Auth` can have code that relies on `UserId` without actually needing to know about your `UserId` (i.e. `Key User`) type. It writes code using `AuthId m`, then you define the type family relationship yourself (e.g. [here](https://github.com/thoughtbot/carnival/blob/master/Foundation.hs#L130)).

When I did:

```haskell
Authenticated userId <- authenticateUser creds

Just user <- runDB $ DB.get userId
```

I got a type error because `userId` is `AuthId m` and `DB.get` requires `UserId`. My extracting the helper that has a concrete `UserId` return type, it made the compiler happy.

But, now that I think about it, it might be the ambiguous `m` that's giving trouble, not the type familiar. I.e. if I fix that to `App` in `Model.User`, it might work without this helper.